### ISSUE TITLE
Filtered Informer is not configured fully. Too many pods are in the informer.

### ISSUE THREAD

**Issue by guillaumerose** [tektoncd/pipeline#4398]
# Expected Behavior

The informer should have only pods with label `app.kubernetes.io/managed-by=tekton-pipelines` or the value of the config `default-managed-by-label-value`

https://github.com/tektoncd/pipeline/blob/main/config/config-defaults.yaml#L52

# Actual Behavior

The informer has too many pods in his memory.

# Steps to Reproduce the Problem

1. Modify the code of the pipeline controller to display the number of pods in the informer
```
go func() {
	for {
		pods, err := podInformer.Lister().List(labels.Everything())
		if err != nil {
			logger.Fatal(err)
		}
		logger.Error(len(pods))
		time.Sleep(time.Second)
	}
}()
```
2. Create a TaskRun
3. Create a pod with a label `app.kubernetes.io/managed-by: not-tekton-at-all`

# Additional Info

The filteredinformerfactory is configured here: https://github.com/tektoncd/pipeline/blob/64aff5edcbb7e6fdf22e1270f60ee37085a8d934/cmd/controller/main.go#L101
The podinformer is configured here: https://github.com/tektoncd/pipeline/blob/64aff5edcbb7e6fdf22e1270f60ee37085a8d934/pkg/reconciler/taskrun/controller.go#L53

**Comment by vdemeester** [comment]
One of the comment in the `config-defaults` for this is:
```
    # default-managed-by-label-value contains the default value given to the
    # "app.kubernetes.io/managed-by" label applied to all Pods created for
    # TaskRuns. If a user's requested TaskRun specifies another value for this
    # label, the user's request supercedes.
```

This is most likely why we are looking at *all* `app.kubernetes.io/managed-by` pods, because users could overwrite it in there task/taskrun to integrate it with their tooling. 
Allowing this or not can be discussed (and is a bit linked to #4366) but as it's currently allowed, I don't see a straightforward way to *fix* this 🙃 . At least we filter out all pods that do not use `app.kubernetes.io/managed-by` labels.

/cc @imjasonh @bobcatfish @afrittoli @sbwsg

**Comment by sbwsg** [comment]
Can we limit the lister to the default label and fork the logic so that a TaskRun with its own managed-by label causes us to skip the lister and query the cluster directly?

**Comment by vdemeester** [comment]
> Can we limit the lister to the default label and fork the logic so that a TaskRun with its own managed-by label causes us to skip the lister and query the cluster directly?

That might be problematic if pipeline is used "transparently" by some higher level tool. Assuming that I have a tool called foo that always setup this `managed-by` label, this would make the controller query the cluster a lot and could become a problem. That said, I do think we may want to have / introduce a label for us that cannot be overriden for this particular purpose. Either we disallow overriding `app.kubernetes.io/manage-by` *or* we come up with our "own" internal one ?

**Comment by imjasonh** [comment]
Can we have two labels, one for Tekton's internal usage, and one for whatever service is built on top of Tekton that might want to hide the Tekton internals (configured by configmap as today)? Then we could have the informer filter by the Tekton-internal label.

Maybe we could use `app.kubernetes.io/created-by: tekton-pipeline-controller`? (see [common labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/))

**Comment by vdemeester** [comment]
> Can we have two labels, one for Tekton's internal usage, and one for whatever service is built on top of Tekton that might want to hide the Tekton internals (configured by configmap as today)? Then we could have the informer filter by the Tekton-internal label.
>
> Maybe we could use `app.kubernetes.io/created-by: tekton-pipeline-controller`? (see [common labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/))

Yes, that was what I was thinking too in my last comment. And I missed `app.kubernetes.io/created-by` label somehow, but I think it's definitely a very good candidate for it. The only thing we need to make sure is to disallow (or prevent) this label to be overriden by the user (by adding it to the `TaskRun` or `PipelineRun` object), either silently or loudly 👼🏼

**Comment by guillaumerose** [comment]
That sounds to be a good idea. I am wondering how it works for the upgrade 🤔 

If a pipeline is done before the upgrade, it's fine.
If a pipeline is running, eg. pods don't have correct labels, after upgrade, the controller won't see them because of the new filter.

Am I correct? Do we need to roll this new label across multiples releases?

**Comment by vdemeester** [comment]
> Am I correct? Do we need to roll this new label across multiples releases?

Good question, we will need to be careful here on how we handle this, but yes we'll have to think of backward compatibility.

**Comment by nikhil-thomas** [comment]
> That sounds to be a good idea. I am wondering how it works for the upgrade 🤔
> 
> If a pipeline is done before the upgrade, it's fine. If a pipeline is running, eg. pods don't have correct labels, after upgrade, the controller won't see them because of the new filter.
> 
> Am I correct? Do we need to roll this new label across multiples releases?

This sounds like a problem for operator to solve.

At present, the operator upgrades TektonPipeline, TektonTriggers etc immediately after an operator controller upgrade. The only check the operator has is version comparission ([pipelines-info configMap](https://github.com/tektoncd/pipeline/blob/76bd56abd0df74d08f4edb60d87cfd32159cc8db/config/config-info.yaml#L29)) of Tektoncd Pipeline (on-cluster vs new release).

It could be straight forward to add a mechanism to delay (requeue) the upgrade if there are active taskRuns. This could be a decent starting point. We could improve this by adding a timeout for this requeue or by implementing a notion of 'maintenance window'.

**Comment by vdemeester** [comment]
## Analysis and Implementation Plan

I've analyzed this issue and have a proposed solution. Here's what I found:

### Root Cause

The filtered informer is configured to filter by label **KEY** (`app.kubernetes.io/managed-by`) but not by **VALUE**. This means the informer caches ALL pods with this label, regardless of whether they're actually managed by Tekton.

**Code locations:**
- `cmd/controller/main.go:109` - Global filter setup
- `pkg/reconciler/taskrun/controller.go:73` - Pod informer configuration
- `pkg/pod/pod.go:568` - Pod label enforcement

### Proposed Solution: Three-Phase Rollout

#### Phase 1: Add Internal Label (Release N)
Add `app.kubernetes.io/created-by: tekton-pipeline-controller` label to all pods:
1. Update `pkg/pod/pod.go:makeLabels()` to add the new label
2. Add validation to reject TaskRuns that try to set this label (prevent user override)
3. No behavior change - just adding the label for future use

**Benefits:** 
- Zero risk - no functionality changes
- Prepares for Phase 2
- All new pods get the internal label

#### Phase 2: Update Filtering (Release N+1)
Switch informer filtering from `managed-by` to `created-by`:
1. Update `cmd/controller/main.go` to filter by `created-by` label
2. Update `pkg/reconciler/taskrun/controller.go` pod informer
3. **This is where we get the memory optimization benefit**

**Benefits:**
- Informer only caches Tekton-created pods
- Compatible with pods from Phase 1
- Users can still customize `managed-by` for their tooling

#### Phase 3: Cleanup (Release N+2 or later)
Optional cleanup once all pods from pre-Phase-1 have completed:
- Remove old filtering code
- Update documentation

### Alternative: Feature Flag

If more control is needed, we could add a feature flag in `config/feature-flags.yaml`:
```yaml
use-created-by-label-filter: "false"  # operators can enable when ready
```

This allows operators to:
- Test in staging environments first
- Enable per-cluster based on readiness
- Rollback if issues arise

### Next Steps

I'm happy to implement Phase 1 (add the `created-by` label + validation) as a first PR. This has zero risk since it doesn't change any behavior, just adds the label for future use.

Thoughts? Which approach would the community prefer - three-phase rollout or feature flag?

---

Full analysis document: https://github.com/tektoncd/pipeline/blob/main/issue-4398-analysis.md (will be in my PR)

**Comment by vdemeester** [comment]
Unfreezing — there's a detailed 3-phase implementation plan posted above. Phase 1 (add `created-by` label) is zero-risk and ready to be picked up.

/remove-lifecycle frozen
